libsolidity/interface/FileReader.h

```diff
@@ -0,0 +1,105 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <boost/filesystem.hpp>
+
+#include <map>
+#include <set>
+
+#include <libsolidity/interface/ImportRemapper.h>
+#include <libsolidity/interface/ReadFile.h>
+
+namespace solidity::frontend
+{
+
+/// FileReader - used for progressively loading source source code.
+///
+/// It is used in solc to load files from CLI parameters, stdin, or from JSON and
+/// also used in the solc language server where solc is a long running process.
+class FileReader
+{
+public:
+	using StringMap = std::map<SourceUnitID, SourceCode>;
+	using PathMap = std::map<SourceUnitID, boost::filesystem::path>;
+	using FileSystemPathSet = std::set<boost::filesystem::path>;
+
+	/// Constructs a FileReader with a base path and a set of allowed directories that
+	/// will be used when requesting files from this file reader instance.
+	explicit FileReader(
+		boost::filesystem::path _basePath = {},
+		FileSystemPathSet _allowedDirectories = {}
+	):
+		m_basePath(std::move(_basePath)),
+		m_allowedDirectories(std::move(_allowedDirectories)),
+		m_sourceCodes(),
+		m_sourceCodeOrigin()
+	{}
+
+	void setBasePath(boost::filesystem::path _path) { m_basePath = std::move(_path); }
+	std::string basePath() const noexcept { return m_basePath.generic_string(); }
```

Comment:
I think this should be symmetrical. It either both accepts and returns `std::string` or uses `boost::filesystem::path` for both.

```suggestion
	boost::filesystem::path basePath() const noexcept { return m_basePath; }
```